src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -0,0 +1,1254 @@
+             << ", remote group snap id: " << snap.id << dendl;
+    C_SaferCond *ctx = new C_SaferCond;
+    create_regular_snapshot(snap.name, snap.id, ctx);
+    ctx->wait();
+    locker.unlock();
+  }
+}
+
+template <typename I>
+void Replayer<I>::create_mirror_snapshot(
+    cls::rbd::GroupSnapshot *snap,
+    const cls::rbd::MirrorSnapshotState &snap_state,
+    std::unique_lock<ceph::mutex> &locker,
+    Context *on_finish) {
+  auto group_snap_id = snap->id;
+  dout(10) << group_snap_id << dendl;
+  ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
+
+  auto itl = std::find_if(
+      m_local_group_snaps.begin(), m_local_group_snaps.end(),
+      [group_snap_id](const cls::rbd::GroupSnapshot &s) {
+      return s.id == group_snap_id;
+      });
+
+  if (itl != m_local_group_snaps.end() &&
+      itl->state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
+    dout(20) << "group snapshot: " << group_snap_id << " already exists"
+             << dendl;
+    locker.unlock();
+    on_finish->complete(0);
+    return;
+  }
+
+  int r;
+  std::set<std::string> mirror_peer_uuids;
+  if (snap_state == cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY_DEMOTED) {
+    std::vector<cls::rbd::MirrorPeer> mirror_peers;
+    r = librbd::cls_client::mirror_peer_list(&m_local_io_ctx, &mirror_peers);
+    if (r < 0) {
+      derr << "failed to list peers: " << cpp_strerror(r) << dendl;
+    }
+
+    for (auto &peer : mirror_peers) {
+      if (peer.mirror_peer_direction == cls::rbd::MIRROR_PEER_DIRECTION_RX) {
+        continue;
+      }
+      mirror_peer_uuids.insert(peer.uuid);
+    }
+  }
+  cls::rbd::GroupSnapshot local_snap =
+  {group_snap_id,
+    cls::rbd::GroupSnapshotNamespaceMirror{
+      snap_state, mirror_peer_uuids, m_remote_mirror_uuid, group_snap_id},
+    {}, cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+  local_snap.name = prepare_non_primary_mirror_snap_name(m_global_group_id,
+      group_snap_id);
+  m_local_group_snaps.push_back(local_snap);
+
+  auto comp = create_rados_callback(
+      new LambdaContext([this, group_snap_id, on_finish](int r) {
+        handle_create_mirror_snapshot(r, group_snap_id, on_finish);
+        }));
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::group_snap_set(&op, local_snap);
+  r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(m_local_group_id), comp, &op);
+  ceph_assert(r == 0);
+  locker.unlock();
+  comp->release();
+
+  // if m_replayer in the ImageReplayer is null this cannot be forwarded.
+  // May be we should retry this setting in the validate_image_snaps_sync_complete().
+  // Same for image_replayer->prune_snapshot(); setting actually!!!!
+  set_image_replayer_limits("", snap);
+}
+
+template <typename I>
+void Replayer<I>::handle_create_mirror_snapshot(
+    int r, const std::string &remote_group_snap_id, Context *on_finish) {
+  dout(10) << remote_group_snap_id << ", r=" << r << dendl;
+
+  on_finish->complete(0);
```

Comment:
This doesn't seem right. Should this be,
```
on_finish->complete(r)
```
Currently,  any error from `group_snap_set` is not being handled.